### PR TITLE
Add rng driver load test

### PR DESCRIPTION
**Pull request by suqinhuang**
1. repeat load/unload 10 times
2. unload during random read

Signed-off-by: Suqin Huang shuang@redhat.com

### CODE REVIEW

**Comment by suqinhuang** [comment]
id: 1354309 1331305

File: qemu/tests/rng_driver_negative.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+        time.sleep(5)
+    except process.CmdError, err:
+        output = err.result.stderr
+        if "(core dumped)" in output:
+            raise exceptions.TestFail(output)
+    finally:
+        error_context.context("Load the driver", logging.info)
+        status, output = session.cmd_status_output(driver_load_cmd)
+        if status != 0:
+            raise exceptions.TestFail("failed to load driver, %s" % output)
```
Comment (CongLi) [inline qemu/tests/rng_driver_negative.py]: If so, I think it's better to open snapshot for the test.

File: qemu/tests/rng_driver_negative.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+from virttest import error_context
+from avocado.utils import process
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    KVM driver load test:
+    1) Log into a guest
+    2) Get module name from guest
```
Comment (CongLi) [inline qemu/tests/rng_driver_negative.py]: Where is this step?

File: qemu/tests/rng_driver_negative.py (hunk 1 above)
Comment (suqinhuang) [inline qemu/tests/rng_driver_negative.py]: good suggestion

**Comment by PyLearner** [comment]
after i applied this patch: https://github.com/autotest/tp-qemu/pull/677 and https://github.com/autotest/tp-qemu/pull/784 , i got some problems when run the case

for smp_4.16384m.repeat1.run_test.Host_RHEL.m7.u3.virtio_rng.rng_random.qcow2.virtio_blk.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.rng_driver_load.load_in_loop , it skipped because of 

ParamNotFound: Mandatory parameter 'driver_id_pattern' is missing

and 
FAIL 2-smp_4.16384m.repeat1.run_test.Host_RHEL.m7.u3.virtio_rng.rng_random.qcow2.virtio_blk.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.rng_driver_load.driver_in_use -> TestFail: random read is not running background

hi @suqinhuang, do i miss some other cfgs ?

**Comment by Zhengtong** [comment]
Tried on power: 

03:28:29 ERROR| SKIP 1-smp_8.4096m.repeat1.run_test.Host_RHEL.m7.u3.virtio_rng.rng_random.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.ppc64le.io-github-autotest-qemu.rng_driver_load.load_in_loop -> ParamNotFound: Mandatory parameter 'driver_id_cmd' is missing. Check your cfg files for typos/mistakes

"driver_id_cmd"is missing , this is different with PyLearner's result, his is "drive_id_pattern"  .

The other result is the same with him.

**Comment by suqinhuang** [comment]
PASS 1-smp_2.8192m.repeat1.run_test.Host_RHEL.m7.u3.virtio_rng.rng_random.qcow2.virtio_blk.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.rng_driver_load.load_in_loop

PASS 2-smp_2.8192m.repeat1.run_test.Host_RHEL.m7.u3.virtio_rng.rng_random.qcow2.virtio_blk.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.rng_driver_load.driver_in_use

**Comment by suqinhuang** [comment]
@Zhengtong @PyLearner patch updated, could you help it again.

Thanks

**Comment by Zhengtong** [comment]
hi @suqinhuang , It seems there are still some cfg options missing in the subtest:"driver_load".  I didn't run pass in my host. Does any more steps I needed to do before running this test scenarios?

02:00:08 ERROR| SKIP 1-smp_16.131072m.repeat1.run_test.Host_RHEL.m7.u3.virtio_rng.rng_random.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.ppc64le.io-github-autotest-qemu.rng_driver_load.load_in_loop -> ParamNotFound: Mandatory parameter 'driver_load_cmd' is missing. Check your cfg files for typos/mistakes

02:02:35 ERROR| FAIL 2-smp_16.131072m.repeat1.run_test.Host_RHEL.m7.u3.virtio_rng.rng_random.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.ppc64le.io-github-autotest-qemu.rng_driver_load.driver_in_use -> TestFail: random read is not running background

**Comment by suqinhuang** [comment]
@Zhengtong because there are two product happen on some hw, I update the method, could you review again

load_in_loop may failed with timeout becasue of the produce bug

**Comment by suqinhuang** [comment]
@aliang123 @Zhengtong updated, thanks

**Comment by aliang123** [comment]
Summary: 
     rng_driver_load.driver_in_use: acked-by Aihua Liang<aliang@redhat.com>
     rng_driver_load.load_in_loop: will check after product bug-1402681 resolved.

Test cmd:
  python ConfigTest.py --guestname=RHEL.7.3 --platform=x86_64 --verbose=yes --testcase=rng_driver_load --rng_type=rng_random
Test Result:
  # python scan_results.py 
CaseName                                 Status          StartTime       EndTime         Result     TimeElapsed
'1-smp_2.4096m.repeat1.run_test.Host_RHEL.m7.u3.virtio_rng.rng_random.qcow2.virtio_blk.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.rng_driver_load.load_in_loop' Finshed 2016-12-09 10:53:40 2016-12-09 10:56:32 ERROR 172.134587
'2-smp_2.4096m.repeat1.run_test.Host_RHEL.m7.u3.virtio_rng.rng_random.qcow2.virtio_blk.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.rng_driver_load.driver_in_use' Finshed 2016-12-09 10:56:32 2016-12-09 11:00:18 PASS 226.565646

**Comment by Zhengtong** [comment]
LGTM
Acked-by Zhengtong Liu(zhengtli@redhat.com)
Passed on ppc64le platform. 

 (1/1) smp_16.131072m.repeat1.run_test.Host_RHEL.m7.u3.virtio_rng.rng_random.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.ppc64le.io-github-autotest-qemu.rng_driver_load.driver_in_use: PASS (239.31 s)

**Comment by PyLearner** [comment]
(1/2) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.virtio_rng.rng_random.qcow2.virtio_blk.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.rng_driver_load.load_in_loop: ERROR (167.47 s)
 (2/2) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.virtio_rng.rng_random.qcow2.virtio_blk.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.rng_driver_load.driver_in_use: PASS (224.41 s)

File: qemu/tests/rng_driver_negative.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    except aexpect.ShellTimeoutError:
+        logging.info("Driver unload timeout")
+    except process.CmdError, err:
+        output = err.result.stderr
+        if "(core dumped)" in output:
+            raise exceptions.TestFail(output)
+    finally:
+        try:
+            process_id = check_bg_test
+            session.cmd("kill -9 %s" % process_id)
```
Comment (xutian) [inline qemu/tests/rng_driver_negative.py]: session.close() here, not need if branch in the end

File: qemu/tests/rng_driver_negative.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        except:
```
Comment (xutian) [inline qemu/tests/rng_driver_negative.py]: better
```
except Exception:
       pass
```

File: qemu/tests/rng_driver_negative.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    2) Read from virtio-rng device
+    3) Unload the device driver
+    4) Check no crash
+    5) Load the device driver
+
+    :param test: QEMU test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment.
+    """
+    def check_bg_test():
```
Comment (xutian) [inline qemu/tests/rng_driver_negative.py]: drop this function

File: qemu/tests/rng_driver_negative.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
+    login_timeout = int(params.get("login_timeout", 360))
+    session = vm.wait_for_login(timeout=login_timeout)
+
+    driver_unload_cmd = params["driver_unload_cmd"]
+
+    error_context.context("Read virtio-rng device in background",
+                          logging.info)
+    session.sendline(params["read_rng_cmd"])
```
Comment (xutian) [inline qemu/tests/rng_driver_negative.py]: output = session.cmd_output_safe("dd if=/dev/random of=/dev/null & echo $!")

File: qemu/tests/rng_driver_negative.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+    if not utils_misc.wait_for(lambda: check_bg_test(), 120, 0, 1):
```
Comment (xutian) [inline qemu/tests/rng_driver_negative.py]: if status != 0:
    raise xxx

File: qemu/tests/rng_driver_negative.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+        raise exceptions.TestFail("random read is not running background")
+
+    try:
+        error_context.context("Unload the driver during random read",
```
Comment (xutian) [inline qemu/tests/rng_driver_negative.py]: move out try

File: qemu/tests/rng_driver_negative.py [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+                              logging.info)
+        output = session.cmd_output(driver_unload_cmd)
+    except aexpect.ShellTimeoutError:
+        logging.info("Driver unload timeout")
```
Comment (xutian) [inline qemu/tests/rng_driver_negative.py]: it's mean pass or fail if timeout?

File: qemu/tests/rng_driver_negative.py [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+    except process.CmdError, err:
+        output = err.result.stderr
+        if "(core dumped)" in output:
+            raise exceptions.TestFail(output)
+    finally:
+        try:
```
Comment (xutian) [inline qemu/tests/rng_driver_negative.py]: need to verify guest dmesg here.

File: qemu/tests/rng_driver_negative.py [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+
+    try:
+        error_context.context("Unload the driver during random read",
+                              logging.info)
+        output = session.cmd_output(driver_unload_cmd)
+    except aexpect.ShellTimeoutError:
+        logging.info("Driver unload timeout")
+    except process.CmdError, err:
+        output = err.result.stderr
+        if "(core dumped)" in output:
```
Comment (xutian) [inline qemu/tests/rng_driver_negative.py]: check core here, just need to logging output

Comment (xutian) [inline qemu/tests/rng_driver_negative.py]: call vm.verify_kernel_crash()

File: qemu/tests/rng_driver_negative.py [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+    session = vm.wait_for_login(timeout=login_timeout)
+
+    driver_unload_cmd = params["driver_unload_cmd"]
+
+    error_context.context("Read virtio-rng device in background",
+                          logging.info)
+    read_rng_cmd = params["read_rng_cmd"]
+    status, pid = session.cmd_status_output(read_rng_cmd)
+
+    if status != 0:
```
Comment (xutian) [inline qemu/tests/rng_driver_negative.py]: you can check it by status, status always 0 after execute your command. please check pid exits

Comment (suqinhuang) [inline qemu/tests/rng_driver_negative.py]: ok, will update it

File: qemu/tests/rng_driver_negative.py [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+
+    driver_unload_cmd = params["driver_unload_cmd"]
+
+    error_context.context("Read virtio-rng device in background",
+                          logging.info)
+    read_rng_cmd = params["read_rng_cmd"]
+    pid = session.cmd_output(read_rng_cmd)
+    pid = pid.split("\n")[1]
+    logging.info("Check if random read process exist")
+    status = session.cmd_status("dir /proc/%s" % pid)
```
Comment (xutian) [inline qemu/tests/rng_driver_negative.py]: how about  ps -p pid?

File: qemu/tests/rng_driver_negative.py [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+    if status != 0:
+        raise exceptions.TestFail("random read is not running background")
+
+    error_context.context("Unload the driver during random read",
+                          logging.info)
+    try:
+        session.cmd(driver_unload_cmd)
+    except aexpect.ShellTimeoutError:
+        pass
+
```
Comment (xutian) [inline qemu/tests/rng_driver_negative.py]: useless blank line.

File: qemu/tests/rng_driver_negative.py (hunk 13 above)
Comment (suqinhuang) [inline qemu/tests/rng_driver_negative.py]: need to analysis output with "ps -p pid", I think check /proc/pid is simple
# ps -p 12345
  PID TTY          TIME CMD

**Comment by suqinhuang** [comment]
PASS 1-smp_16.16384m.repeat1.run_test.Host_RHEL.m7.u3.virtio_rng.rng_random.qcow2.virtio_blk.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.rng_driver_load.driver_in_use

**Comment by xutian** [comment]
merged, thanks.

**Pull request merged**